InternalDocs/stackrefs.md

```diff
@@ -0,0 +1,79 @@
+# Stack references (`_PyStackRef`)
+
+Stack references are the interpreter's tagged representation of values on the evaluation stack.
+They carry metadata to track ownership and support optimizations such as tagged small ints.
+
+## Shape and tagging
+
+- A `_PyStackRef` is a tagged pointer-sized value (see `Include/internal/pycore_stackref.h`).
+- Tag bits distinguish three cases:
+  - `Py_TAG_REFCNT` clear - reference count lives on the pointed-to object.
+  - `Py_TAG_REFCNT` set - ownership is "borrowed" (no refcount to drop on close) or the object is immortal.
+  - `Py_INT_TAG` - tagged small integer stored directly in the stackref (no heap allocation).
+- Special constants: `PyStackRef_NULL`, `PyStackRef_ERROR`, and embedded `None`/`True`/`False`.
+
+In GIL builds, most objects carry their refcount; tagged borrowed refs skip decref on close. In free
+threading builds, the tag is also used to mark deferred refcounted objects so the GC can see them and
+to avoid refcount contention for short-lived stack values.
+
+## Converting to and from PyObject*
+
+Three conversions control ownership:
+
+- `PyStackRef_FromPyObjectNew(obj)` - create a new reference (INCREF if mortal).
+- `PyStackRef_FromPyObjectSteal(obj)` - take over ownership without changing the count unless the
+  object is immortal.
+- `PyStackRef_FromPyObjectBorrow(obj)` - create a borrowed stackref (never decref on close).
+
+The `obj` argument must not be `NULL`.
+
+Going back to `PyObject*` mirrors this:
+
+- `PyStackRef_AsPyObjectBorrow(ref)` - borrow the underlying pointer
+- `PyStackRef_AsPyObjectSteal(ref)` - transfer ownership from the stackref
```

Comment:
If `ref` is borrowed/deferred this will create a new owning reference. Not sure it's worth documenting, but I figured I'd call it out since this is not exactly the dual of `PyStackRef_FromPyObjectSteal`.